x-pack/apm-server/sampling/eventstorage/storage.go

```diff
@@ -185,30 +172,70 @@ func (rw *ReadWriter) WriteTraceEvent(traceID string, id string, event *modelpb.
 
 func (rw *ReadWriter) writeEntry(e *badger.Entry, opts WriterOpts) error {
 	rw.pendingWrites++
-	err := rw.txn.SetEntry(e.WithTTL(opts.TTL))
-	// Attempt to flush if there are 200 or more uncommitted writes.
-	// This ensures calls to ReadTraceEvents are not slowed down;
-	// ReadTraceEvents uses an iterator, which must sort all keys
-	// of uncommitted writes.
-	// The 200 value yielded a good balance between read and write speed:
-	// https://github.com/elastic/apm-server/pull/8407#issuecomment-1162994643
+	entrySize := estimateSize(e)
+	lsm, vlog := rw.s.db.Size()
+
+	// there are multiple ReadWriters writing to the same storage so add
+	// the entry size and consider the new value to avoid TOCTOU issues.
+	pendingSize := rw.s.pendingSize.Add(entrySize)
+	rw.pendingSize += entrySize
+
+	if current := pendingSize + lsm + vlog; opts.StorageLimitInBytes != 0 && current >= opts.StorageLimitInBytes {
+		// flush what we currently have and discard the current entry
+		if err := rw.Flush(); err != nil {
+			return err
+		}
+		return fmt.Errorf("%w (current: %d, limit: %d)", ErrLimitReached, current, opts.StorageLimitInBytes)
+	}
+
 	if rw.pendingWrites >= 200 {
-		if err := rw.Flush(opts.StorageLimitInBytes); err != nil {
+		// Attempt to flush if there are 200 or more uncommitted writes.
+		// This ensures calls to ReadTraceEvents are not slowed down;
+		// ReadTraceEvents uses an iterator, which must sort all keys
+		// of uncommitted writes.
+		// The 200 value yielded a good balance between read and write speed:
+		// https://github.com/elastic/apm-server/pull/8407#issuecomment-1162994643
+		if err := rw.Flush(); err != nil {
 			return err
 		}
+
+		// the current ReadWriter flushed the transaction and reset the pendingSize so add
+		// the entrySize again.
+		rw.pendingSize += entrySize
+		rw.s.pendingSize.Add(entrySize)
```

Comment:
this is possible to race & overshoot storage limit. Imagine this:

normal scenario:
rw1: line 198 flush, rw1.s.pendingSize.Add(-rw1.pendingSize)
rw1: line 204-205 pendingSize increase
rw2: line 175-183, failed size check, flush immediately and reject current entry

race scenario that leads to overshooting the limit:
rw1: line 198 flush, rw1.s.pendingSize.Add(-rw1.pendingSize)
rw2: line 175-183, passes size check
rw2: line 208-214, setEntry successfully
rw1: line 204-205 pendingSize increase
rw1: line 208-214, setEntry successfully